lib/ansible/modules/cloud/amazon/ec2_snapshot.py

```diff
@@ -231,24 +251,66 @@ def create_snapshot(module, ec2, state=None, description=None, wait=None,
         if snapshot is None:
-            snapshot = ec2.create_snapshot(volume_id, description=description)
-            changed = True
+            if not module.check_mode:
+                snapshot = ec2.create_snapshot(volume_id, description=description)
+                changed = True
+            else:
+                module.exit_json(changed=True)
         if wait:
             if not _create_with_wait(snapshot, wait_timeout):
                 module.fail_json(msg='Timed out while creating snapshot.')
-        if snapshot_tags:
-            for k, v in snapshot_tags.items():
+
+        tags_to_add, tags_to_remove = existing_to_desired(snapshot.tags.items(), snapshot_tags.items())
+        for (k, v) in tags_to_add:
+            if not module.check_mode:
                 snapshot.add_tag(k, v)
+            changed = True
+        for (k, v) in tags_to_remove:
+            if not module.check_mode:
+                snapshot.remove_tag(k, value=v)
+            changed = True
+
+        if module.check_mode:
+            tags = snapshot_tags.copy()
+        else:
+            tags = snapshot.tags.copy()
+
+        permissions = ec2.get_snapshot_attribute(snapshot.id)
+
+        users_to_add, users_to_remove = existing_to_desired(permissions.attrs.get('user_ids', []),
+                                                            [str(user_id) for user_id in
+                                                             create_volume_permissions.get('user_ids', [])])
+        groups_to_add, groups_to_remove = existing_to_desired(permissions.attrs.get('groups', []),
+                                                              create_volume_permissions.get('groups', []))
+
+        if users_to_add or groups_to_add:
+            if not module.check_mode:
+                ec2.modify_snapshot_attribute(snapshot.id, user_ids=users_to_add, groups=groups_to_add)
+                permissions = ec2.get_snapshot_attribute(snapshot.id)
+            else:
+                permissions = create_volume_permissions
+            changed = True
+        if users_to_remove or groups_to_remove:
+            if not module.check_mode:
+                ec2.modify_snapshot_attribute(snapshot.id, operation='remove',
+                                              user_ids=users_to_remove, groups=groups_to_remove)
+                permissions = ec2.get_snapshot_attribute(snapshot.id)
+            else:
+                permissions = create_volume_permissions
+            changed = True
     except boto.exception.BotoServerError as e:
         module.fail_json(msg="%s: %s" % (e.error_code, e.error_message))
 
     module.exit_json(changed=changed,
                      snapshot_id=snapshot.id,
                      volume_id=snapshot.volume_id,
                      volume_size=snapshot.volume_size,
-                     tags=snapshot.tags.copy())
+                     tags=tags,
+                     permissions=permissions.attrs)
```

Comment:
This will fail in check mode if groups are being added or removed since permissions won't have attrs. I also notice there seems to be a lag with updating removals - I can remove a user and it will show the results as being changed but the user id remains. Would it make sense to always set permissions to create_volume_permissions if a change is being made? Although it would be easier to see things going wrong if a call is actually being made to get the latest state...